<font size="7"><span style="color: #33D6CC;">Exploración genérica del dataset</span></font>

In [2]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt 
import warnings
warnings.filterwarnings(action='ignore',category=FutureWarning)

In [3]:
df = pd.read_csv('tenis_bruto.csv')
df

C:\Users\albam\AppData\Local\Temp\ipykernel_23428\584941035.py:1: DtypeWarning: Columns (8,16) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('tenis_bruto.csv')


tourney_id                  tourney_name  \
0                              1968-2029                        Dublin   
1                              1968-2029                        Dublin   
2                              1968-2029                        Dublin   
3                              1968-2029                        Dublin   
4                              1968-2029                        Dublin   
...                                  ...                           ...   
193332  2024-M-DC-2024-WG2-PO-URU-MDA-01  Davis Cup WG2 PO: URU vs MDA   
193333  2024-M-DC-2024-WG2-PO-VIE-RSA-01  Davis Cup WG2 PO: VIE vs RSA   
193334  2024-M-DC-2024-WG2-PO-VIE-RSA-01  Davis Cup WG2 PO: VIE vs RSA   
193335  2024-M-DC-2024-WG2-PO-VIE-RSA-01  Davis Cup WG2 PO: VIE vs RSA   
193336  2024-M-DC-2024-WG2-PO-VIE-RSA-01  Davis Cup WG2 PO: VIE vs RSA   

       surface  draw_size tourney_level  tourney_date  match_num  winner_id  \
0        Grass       32.0             A      19680708        270     112411   
1        Grass       32.0             A      19680708        271     126914   
2        Grass       32.0             A      19680708        272     209523   
3        Grass       32.0             A      19680708        273     100084   
4        Grass       32.0             A      19680708        274     100132   
...        ...        ...           ...           ...        ...        ...   
193332    Clay        4.0             D      20240203          5     212051   
193333    Hard        4.0             D      20240202          1     122533   
193334    Hard        4.0             D      20240202          2     144748   
193335    Hard        4.0             D      20240202          4     122533   
193336    Hard        4.0             D      20240202          5     202475   

       winner_seed winner_entry  ... l_1stIn l_1stWon  l_2ndWon l_SvGms  \
0              NaN          NaN  ...     NaN      NaN       NaN     NaN   
1              NaN          NaN  ...     NaN      NaN       NaN     NaN   
2              NaN          NaN  ...     NaN      NaN       NaN     NaN   
3              NaN          NaN  ...     NaN      NaN       NaN     NaN   
4              NaN          NaN  ...     NaN      NaN       NaN     NaN   
...            ...          ...  ...     ...      ...       ...     ...   
193332         NaN          NaN  ...    30.0     17.0       7.0     6.0   
193333         NaN          NaN  ...    41.0     25.0       6.0     9.0   
193334         NaN          NaN  ...    51.0     25.0       7.0    11.0   
193335         NaN          NaN  ...    51.0     32.0      17.0    14.0   
193336         NaN          NaN  ...    53.0     25.0       8.0     8.0   

        l_bpSaved  l_bpFaced winner_rank winner_rank_points loser_rank  \
0             NaN        NaN         NaN                NaN        NaN   
1             NaN        NaN         NaN                NaN        NaN   
2             NaN        NaN         NaN                NaN        NaN   
3             NaN        NaN         NaN                NaN        NaN   
4             NaN        NaN         NaN                NaN        NaN   
...           ...        ...         ...                ...        ...   
193332        8.0       14.0      1109.0                8.0      740.0   
193333        1.0        4.0       554.0               67.0      748.0   
193334        5.0       12.0       416.0              109.0        NaN   
193335        5.0        9.0       554.0               67.0      416.0   
193336        9.0       14.0       748.0               32.0        NaN   

       loser_rank_points  
0                    NaN  
1                    NaN  
2                    NaN  
3                    NaN  
4                    NaN  
...                  ...  
193332              34.0  
193333              32.0  
193334               NaN  
193335             109.0  
193336               NaN  

[193337 rows x 49 columns]

In [4]:
df.columns

Index(['tourney_id', 'tourney_name', 'surface', 'draw_size', 'tourney_level',
       'tourney_date', 'match_num', 'winner_id', 'winner_seed', 'winner_entry',
       'winner_name', 'winner_hand', 'winner_ht', 'winner_ioc', 'winner_age',
       'loser_id', 'loser_seed', 'loser_entry', 'loser_name', 'loser_hand',
       'loser_ht', 'loser_ioc', 'loser_age', 'score', 'best_of', 'round',
       'minutes', 'w_ace', 'w_df', 'w_svpt', 'w_1stIn', 'w_1stWon', 'w_2ndWon',
       'w_SvGms', 'w_bpSaved', 'w_bpFaced', 'l_ace', 'l_df', 'l_svpt',
       'l_1stIn', 'l_1stWon', 'l_2ndWon', 'l_SvGms', 'l_bpSaved', 'l_bpFaced',
       'winner_rank', 'winner_rank_points', 'loser_rank', 'loser_rank_points'],
      dtype='object')

## Campo de datos:

- **tourney_id** : Identificador único para cada torneo
- **tourney_name**: Nombre del torneo
- **surface**: Tipo de superficie de la pista:
  - *Grass*: Césped
  - *Clay*: Tierra.
  - *Carpet*: Moqueta
  - *Hard*: Pista dura
- **draw_size**: tamaño del cuadro: número de jugadores en él, a menudo redondeado a la potencia de 2 más cercana
- **tourney_level**: nivel del torneo:
  - *A*: Otros torneos del circuito
  - *M* : Master 1000
  - *G*: Grand Slam
  - *F*: Finales del circuito y otros torneos de final de temporada
  - *D*: Copa Davis
- **tourney_date**: Fecha del torneo.Normalmente el lunes de la semana del torneo
- **match_num**: Identificador de partido
- **winner_id**: id del ganador del partido
- **winner_seed**: Número de cabeza de serie.Cuánto mas bajo mejor.Se aplica a la planificación del torneo para que los jugadores de más nivel no se enfrenten entre sí en las primeras rondas 
- **winner_entry**:Valores str de como entró el jugador al torneo: *'Q', 'LL', 'WC', 'PR', 'SE', 'ALT'*
- **winner_name**: nombre del ganador del partido
- **winner_hand**: mano que utiliza el jugador ganador del partido
- **winner_ht**: altura del ganador del partido
- **winner_ioc**: Nacionalidad del ganador del partido
- **winner_age**: Edad del ganador del partido
- **loser_id**: id del perdedor del partido
- **loser_seed**: Número de cabeza de serie.
- **loser_entry**: Valores str de como entró el jugador al torneo: *'Q', 'LL', 'WC', 'PR', 'SE', 'ALT'*
- **loser_name**:nombre del perdedor del partido
- **loser_hand**:mano que utiliza el jugador perdedor del partido
- **loser_ht**: Altura del perdedor del partido
- **loser_ioc**: Nacionalidad del perdedor del partido
- **loser_age**: Edad del perdedor del partido
- **score**: Marcador final del partido
- **best_of**: Si el torneo es al mejor de *3* sets o de *5* sets (2 ganados o 3 ganados) 
- **round**: ronda del torneo 
- **minutes**: Duración del partido en minutos. TARGET
- **w_ace**: Número de aces del ganador
- **w_df**: Número de dobles faltas cometidas por el ganador 
- **w_svpt**: Número de primeros servicios efectuados por el ganador
- **w_1stIn**: Número de primeros servicios válidos del ganador 
- **w_1stWon**: Número de primeros servicios ganados por el ganador
- **w_2ndWon**: Número de segundos servicios ganados por el ganador 
- **w_SvGms**: Número de juegos ganados por el ganador 
- **w_bpSaved**: Número de puntos de break salvados por el ganador 
- **w_bpFaced**: Número de puntos de break afrontados por el ganador 
- **l_ace**: Numero de aces del perdedor 
- **l_df**: Número de dobles faltas cometidas por el perdedor 
- **l_svpt**:Número de primeros servicios efectuados por el perdedor 
- **l_1stIn**: Número de primeros servicios válidos del perdedor  
- **l_1stWon**: Número de primeros servicios ganados por el perdedor
- **l_2ndWon**: Número de segundos servicios ganados por el perdedor 
- **l_SvGms**: Número de juegos ganados por el ganador
- **l_bpSaved**: Número de puntos de break salvados por el perdedor
- **l_bpFaced**: Número de puntos de break afrontados por el perdedor
- **winner_rank**: Número de ranking del ganador 
- **winner_rank_points**: Número de puntos de ranking del ganador
- **loser_rank**: Número de ranking del perdedor 
- **loser_rank_points**: Número de puntos de ranking del perdedor

49 Columnas, 193337 Filas

In [5]:
df.surface.unique()

array(['Grass', 'Clay', 'Carpet', 'Hard', nan], dtype=object)

In [6]:
df[df['surface']=='Carpet'] # en 2009 la prohibieron y cambiaron todas por pista dura

tourney_id                 tourney_name  \
129                               1968-401             Philadelphia WCT   
130                               1968-401             Philadelphia WCT   
131                               1968-401             Philadelphia WCT   
132                               1968-401             Philadelphia WCT   
133                               1968-401             Philadelphia WCT   
...                                    ...                          ...   
173523  2017-M-DC-2017-G2-EPA-M-FIN-GEO-01  Davis Cup G2 R1: FIN vs GEO   
173535  2017-M-DC-2017-G2-EPA-M-MAR-DEN-01  Davis Cup G2 R1: MAR vs DEN   
173536  2017-M-DC-2017-G2-EPA-M-MAR-DEN-01  Davis Cup G2 R1: MAR vs DEN   
173537  2017-M-DC-2017-G2-EPA-M-MAR-DEN-01  Davis Cup G2 R1: MAR vs DEN   
173538  2017-M-DC-2017-G2-EPA-M-MAR-DEN-01  Davis Cup G2 R1: MAR vs DEN   

       surface  draw_size tourney_level  tourney_date  match_num  winner_id  \
129     Carpet       16.0             A      19680208          1     100099   
130     Carpet       16.0             A      19680208          2     100128   
131     Carpet       16.0             A      19680208          3     100074   
132     Carpet       16.0             A      19680208          4     100011   
133     Carpet       16.0             A      19680208          5     100081   
...        ...        ...           ...           ...        ...        ...   
173523  Carpet        4.0             D      20170203          5     106106   
173535  Carpet        4.0             D      20170203          1     104233   
173536  Carpet        4.0             D      20170203          2     105149   
173537  Carpet        4.0             D      20170203          4     104233   
173538  Carpet        4.0             D      20170203          5     126526   

       winner_seed winner_entry  ... l_1stIn l_1stWon  l_2ndWon l_SvGms  \
129            NaN          NaN  ...     NaN      NaN       NaN     NaN   
130            NaN          NaN  ...     NaN      NaN       NaN     NaN   
131            3.0          NaN  ...     NaN      NaN       NaN     NaN   
132            NaN          NaN  ...     NaN      NaN       NaN     NaN   
133            1.0          NaN  ...     NaN      NaN       NaN     NaN   
...            ...          ...  ...     ...      ...       ...     ...   
173523         NaN          NaN  ...     1.0      0.0       0.0    26.0   
173535         NaN          NaN  ...    43.0     28.0      14.0    12.0   
173536         NaN          NaN  ...    82.0     54.0      24.0    18.0   
173537         NaN          NaN  ...    61.0     46.0      23.0    18.0   
173538         NaN          NaN  ...    37.0     24.0       7.0    10.0   

        l_bpSaved  l_bpFaced winner_rank winner_rank_points loser_rank  \
129           NaN        NaN         NaN                NaN        NaN   
130           NaN        NaN         NaN                NaN        NaN   
131           NaN        NaN         NaN                NaN        NaN   
132           NaN        NaN         NaN                NaN        NaN   
133           NaN        NaN         NaN                NaN        NaN   
...           ...        ...         ...                ...        ...   
173523        0.0        0.0       819.0               21.0      745.0   
173535        3.0        8.0       408.0              104.0      861.0   
173536        9.0       15.0       682.0               36.0        NaN   
173537        4.0        8.0       408.0              104.0      682.0   
173538        4.0       10.0         NaN                NaN     1252.0   

       loser_rank_points  
129                  NaN  
130                  NaN  
131                  NaN  
132                  NaN  
133                  NaN  
...                  ...  
173523              28.0  
173535              18.0  
173536               NaN  
173537              36.0  
173538               5.0  

[20900 rows x 49 columns]

In [7]:
df['tourney_level'].unique()

array(['A', 'M', 'G', 'D', 'F'], dtype=object)

In [8]:
df.winner_entry.unique()

array([nan, 'Q', 'LL', 'WC', 'PR', 'SE', 'ALT', 'Alt'], dtype=object)

## Análisis genérico

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193337 entries, 0 to 193336
Data columns (total 49 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   tourney_id          193337 non-null  object 
 1   tourney_name        193337 non-null  object 
 2   surface             190347 non-null  object 
 3   draw_size           192502 non-null  float64
 4   tourney_level       193337 non-null  object 
 5   tourney_date        193337 non-null  int64  
 6   match_num           193337 non-null  int64  
 7   winner_id           193337 non-null  int64  
 8   winner_seed         71544 non-null   object 
 9   winner_entry        16943 non-null   object 
 10  winner_name         193337 non-null  object 
 11  winner_hand         193327 non-null  object 
 12  winner_ht           176236 non-null  float64
 13  winner_ioc          193329 non-null  object 
 14  winner_age          192026 non-null  float64
 15  loser_id            193337 non-nul

**Comentario**: round formato str

In [10]:
df.describe()

draw_size  tourney_date      match_num      winner_id  \
count  192502.000000  1.933370e+05  193337.000000  193337.000000   
mean       53.184497  1.993934e+07      80.579191  104863.644414   
std        36.755651  1.596434e+05     112.614086   14758.641896   
min         2.000000  1.967123e+07       1.000000  100001.000000   
25%        32.000000  1.980061e+07      11.000000  100428.000000   
50%        32.000000  1.993073e+07      26.000000  101736.000000   
75%        64.000000  2.007071e+07     100.000000  104026.000000   
max       128.000000  2.024052e+07    1701.000000  212721.000000   

           winner_ht     winner_age      loser_id       loser_ht  \
count  176236.000000  192026.000000  193337.00000  163431.000000   
mean      184.512636      25.675078  106029.47141     184.279941   
std         6.673546       4.054686   17484.75218       6.658337   
min       160.000000      14.300000  100001.00000     160.000000   
25%       180.000000      22.700000  100529.00000     180.000000   
50%       185.000000      25.300000  101926.00000     185.000000   
75%       188.000000      28.200000  104417.00000     188.000000   
max       211.000000      58.700000  212722.00000     211.000000   

           loser_age        best_of  ...       l_1stIn      l_1stWon  \
count  188680.000000  193337.000000  ...  96442.000000  96442.000000   
mean       25.729239       3.435861  ...     48.091495     31.963564   
std         4.201515       0.826583  ...     19.357141     14.434812   
min        14.200000       1.000000  ...      0.000000      0.000000   
25%        22.700000       3.000000  ...     34.000000     22.000000   
50%        25.400000       3.000000  ...     45.000000     30.000000   
75%        28.400000       3.000000  ...     58.000000     40.000000   
max        63.600000       5.000000  ...    328.000000    284.000000   

           l_2ndWon       l_SvGms     l_bpSaved     l_bpFaced    winner_rank  \
count  96442.000000  96443.000000  96442.000000  96442.000000  157569.000000   
mean      14.954812     12.199838      4.805189      8.721937      75.295509   
std        7.197164      4.126408      3.273632      4.129425     121.181395   
min        0.000000      0.000000     -6.000000      0.000000       1.000000   
25%       10.000000      9.000000      2.000000      6.000000      17.000000   
50%       14.000000     11.000000      4.000000      8.000000      45.000000   
75%       19.000000     15.000000      7.000000     11.000000      86.000000   
max      101.000000     91.000000     28.000000     38.000000    2101.000000   

       winner_rank_points     loser_rank  loser_rank_points  
count       110346.000000  149182.000000      108702.000000  
mean          1381.821425     112.896308         869.598701  
std           1734.519305     162.708561         998.483982  
min              0.000000       1.000000           0.000000  
25%            497.000000      37.000000         367.000000  
50%            852.000000      70.000000         638.000000  
75%           1547.000000     118.000000        1019.000000  
max          16950.000000    2159.000000       16950.000000  

[8 rows x 33 columns]

**Comentario**: 16 Columnas categóricas, 33 numéricas. Muchos nulos.

In [11]:
df.isnull().mean()*100 #porcentaje de nulos

tourney_id             0.000000
tourney_name           0.000000
surface                1.546522
draw_size              0.431888
tourney_level          0.000000
tourney_date           0.000000
match_num              0.000000
winner_id              0.000000
winner_seed           62.995185
winner_entry          91.236546
winner_name            0.000000
winner_hand            0.005172
winner_ht              8.845177
winner_ioc             0.004138
winner_age             0.678091
loser_id               0.000000
loser_seed            81.133461
loser_entry           85.141489
loser_name             0.000000
loser_hand             0.025344
loser_ht              15.468327
loser_ioc              0.037241
loser_age              2.408747
score                  0.004655
best_of                0.000000
round                  0.000000
minutes               51.544712
w_ace                 50.116636
w_df                  50.117153
w_svpt                50.117153
w_1stIn               50.117153
w_1stWon

**Comentario**: Porcentaje igual en varias columnas lo que seguramente signifique que corresponden a las mismas filas. Las variables '_entry' y '_seed' demasiados nulos asique mejor borrar.


**Comentario**: Mirando el csv de tenis_bruto: la mayoría de nulos están en los años del '68 al '90.